_templates/l10n.php

```diff
@@ -35,334 +37,368 @@ function list_langs() {
         'zh_CN' => '简体中文',
         'zh_TW' => '繁體中文',
     );
-}
 
-function lang_dir($lang) {
-    return dirname(__FILE__).'/../lang/'.$lang;
-}
+    protected $lang = 'en';
 
-function is_lang($lang) {
-    if (!is_string($lang)) {
-        return false;
-    }
-    if (!preg_match('#^[a-z]{2}(_([A-Z]{2}|[A-Z][a-z]+))?$#', $lang)) {
-        return false;
+    protected $domain = null;
+
+    public function __construct($lang = null) {
+        if (empty($lang)) {
+            $lang = $this->get_page_lang();
+        }
+        $this->lang = $lang;
     }
-    if ($lang == 'en') {
-        return true;
+
+    public function init() {
+        global $sitewide, $page; // Global site variables
+
+        if (defined('HTML_I18N')) {
+            return;
+        }
+
+        // Redirect the user if we are translating the page
+        if ((isset($_GET['lang']) || isset($_COOKIE['language'])) 
+            && (isset($_GET['lang']) ? $_GET['lang'] : 'en') != $this->lang
+            && $this->lang != 'en') {
+
+            $url = $sitewide['root'];
+            $url .= $this->lang.$page['path'];
+            $url = '/'.ltrim($url, '/'); // Make sure there is a / at the begining
+            header('Location: '.$url);
+            exit();
+        }
+
+        // Set cookie if the user has chosen the current language
+        if (isset($_GET['lang'])) {
+            setcookie('language', $this->lang,  time() + 60*60*24*30, '/'); // 30 days
+        }
     }
 
-    return is_dir(lang_dir($lang));
-}
+    public function list_langs() {
+        return $this->available_langs;
+    }
 
-function user_lang() {
-    if (isset($_COOKIE['language']) && is_lang($_COOKIE['language'])) {
-        return $_COOKIE['language'];
+    public function lang_dir($lang) {
+        return self::LANG_DIR.'/'.$lang;
     }
 
-    if (empty($_SERVER['HTTP_ACCEPT_LANGUAGE'])) {
-        return null;
+    public function is_lang($lang) {
+        if (!is_string($lang)) {
+            return false;
+        }
+        if (!preg_match('#^[a-z]{2}(_([A-Z]{2}|[A-Z][a-z]+))?$#', $lang)) {
```

Comment:
Available languages are a list displayed to the user in the footer. But it's better not to rely on it to be able to test incomplete translations.